auth/token/token_provider.go

```diff
@@ -0,0 +1,167 @@
+	return string(rt)
+}
+
+// GrantType authorization model
+type GrantType string
+
+// define authorization model
+const (
+	AuthorizationCode   GrantType = "authorization_code"
+	PasswordCredentials GrantType = "password"
+	ClientCredentials   GrantType = "client_credentials"
+	Refreshing          GrantType = "refresh_token"
+	Implicit            GrantType = "__implicit"
+)
+
+func (gt GrantType) String() string {
+	if gt == AuthorizationCode ||
+		gt == PasswordCredentials ||
+		gt == ClientCredentials ||
+		gt == Refreshing {
+		return string(gt)
+	}
+	return ""
+}
+
+// LocalProvider implements the Provider interface and
+// provides mechanics for generating tokens for a selected profile
+type LocalProvider struct {
+	profiles profile.Store
+	keys     key.Store
+}
+
+// NewProvider instantiates a new LocalProvider
+func NewProvider(p profile.Store, k key.Store) (*LocalProvider, error) {
+	return &LocalProvider{
+		profiles: p,
+		keys:     k,
+	}, nil
+}
+
+// compile-time assertion that LocalProvider is a token.Provider
+var _ Provider = (*LocalProvider)(nil)
+
+// Token handles the OAuth token flow
+func (p *LocalProvider) Token(ctx context.Context, req *Request) (*Response, error) {
+	log.Debugf("token.Provider got request: %+v", req)
+	resp := &Response{TokenType: "jwt", ExpiresIn: int64(AccessTokenTTL.Seconds())}
+	switch req.GrantType {
+	case PasswordCredentials:
+		if req.Username == "" {
+			return nil, ErrInvalidCredentials
+		}
+		// TODO(arqu): this only selects the first returned profile for a given peername.
+		// ideally we would use the profile.ID to fetch the exact profile
+		// or otherwise validate the signatures
+		pros, err := p.profiles.ProfilesForUsername(req.Username)
+		if err != nil {
+			log.Debugf("token.Provider failed to fetch profiles: %q", err.Error())
+			return nil, ErrInvalidRequest
+		}
+		if len(pros) == 0 {
+			log.Debugf("token.Provider no matching profiles found")
+			return nil, ErrNotFound
+		}
+		pro := pros[0]
+		if pro.PrivKey == nil {
+			log.Debugf("token.Provider private key is nil")
+			return nil, ErrInvalidCredentials
+		}
+		accessToken, err := NewPrivKeyAuthToken(pro.PrivKey, pro.ID.String(), AccessTokenTTL)
+		if err != nil {
+			log.Debugf("token.Provider failed to generate access token: %q", err.Error())
+			return nil, ErrInvalidRequest
+		}
+		refreshToken, err := NewPrivKeyAuthToken(pro.PrivKey, pro.ID.String(), RefreshTokenTTL)
+		if err != nil {
+			log.Debugf("token.Provider failed to generate refresh token: %q", err.Error())
+			return nil, ErrInvalidRequest
+		}
+		resp.AccessToken = accessToken
+		resp.RefreshToken = refreshToken
+	case Refreshing:
+		tok, err := ParseAuthToken(req.RefreshToken, p.keys)
+		if err != nil {
+			log.Debugf("token.Provider error parsing refresh token: %q", err.Error())
+			return nil, ErrInvalidRequest
+		}
+
+		if claims, ok := tok.Claims.(*Claims); ok {
+			pro, err := p.profiles.GetProfile(profile.IDB58DecodeOrEmpty(claims.ProfileID))
```

Comment:
Since we already have a bunch of error handling here, and the `claims.ProfileID` is coming from external user input, and passing an empty string to `profiles.GetProfile` will mask decoding errors, it would be better here to use `profile.IDB58Decode` and return an error rather than using the `...OrEmpty` version.

Reply:
did the same fix on `lib/lib.activeProfile`